[Fix #21022] Improve performance by reducing re-renders SidebarLinks

hannojg opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

Improve the performance by reducing unnecessary re-renders of the `SidebarLinks` component.

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/app/issues/21022
PROPOSAL: https://github.com/expensify/app/issues/21022#issuecomment-1600631095


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console
- [ ] Open a report, verify that the highlight of the items still work
- [ ] Send a message from user A to user B and verify that the reports are resorting accordingly

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

n/a

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console
- [ ] Open a report, verify that the highlight of the items still work
- [ ] Send a message from user A to user B and verify that the reports are resorting accordingly

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

<img width="1512" alt="browser" src="https://github.com/Expensify/App/assets/16821682/d28acaaf-74f9-46db-8146-33ea49ce387f">


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

![mWeb-Chrome](https://github.com/Expensify/App/assets/16821682/d6718be3-f113-42a5-acc2-8b6b876456bb)


</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

![ios_safari](https://github.com/Expensify/App/assets/16821682/ed38af54-a682-4ddd-affb-2bf2e56954a7)


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

<img width="1312" alt="desktop" src="https://github.com/Expensify/App/assets/16821682/1f353f08-c501-42c8-a78b-880192860e09">


</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

![ios](https://github.com/Expensify/App/assets/16821682/e2d57801-1050-4fad-a6bb-21bfc5eda513)


</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

![android](https://github.com/Expensify/App/assets/16821682/ad10246b-c92e-43f1-90aa-f36ea21b4e13)


</details>

hannojg committed:
Connect OptionRowLHN manually

hannojg committed:
don't rerender SidebarLinks just because modal opens

hannojg committed:
Remove currentReportId from shouldReportBeInOptionList

There isn't any case where the current active report wouldn't be in the list of reports

hannojg committed:
add active report functionality back

hannojg committed:
optimize optionRowLHN to not depend on current report id

venture1981 commented on src/libs/ReportUtils.js:
```diff
@@ -1891,14 +1891,13 @@ function canSeeDefaultRoom(report, policies, betas) {
  * filter out the majority of reports before filtering out very specific minority of reports.
  *
  * @param {Object} report
- * @param {String} currentReportId
```
this `currentReportId` is unnecessary variable in the code?

venture1981 reviewed (commented):
@hannojg 
How about css optimization?

hannojg committed:
fix

hannojg commented on src/components/LHNOptionsList/OptionRowLHN.js:
```diff
@@ -254,4 +261,34 @@ OptionRowLHN.propTypes = propTypes;
+OptionRowIsFocusedSupport.defaultProps = defaultProps;
+OptionRowIsFocusedSupport.displayName = 'OptionRowIsFocusedSupport';
+
+// TODO: Note on mobile we could skip this HOC
```
No, sorry, I wasn't pushing the latest changes. On mobile we can't skip this HOC, so never mind

hannojg commented on src/libs/SidebarUtils.js:
```diff
@@ -20,6 +20,7 @@ import * as UserUtils from './UserUtils';
 // Session also can remain stale because the only way for the current user to change is to sign out and sign in, which would clear out all the Onyx
 // data anyway and cause SidebarLinks to rerender.
 
+// TODO: get rid of this shit
```
that shouldn't be here, sorry

hannojg commented on src/pages/home/sidebar/SidebarLinks.js:
```diff
@@ -274,8 +281,6 @@ const chatReportSelector = (report) =>
         errorFields: {
             addWorkspaceRoom: report.errorFields && report.errorFields.addWorkspaceRoom,
         },
-        lastReadTime: report.lastReadTime,
-        lastMentionedTime: report.lastMentionedTime,
```
Yes, this is the main performance optimisation as described in the proposal

hannojg commented:
> How about css optimization?

Sorry, what do you mean by that @venture1981

hannojg commented:
@0xmiroslav I think I did it correctly, no?

![Screenshot 2023-06-26 at 10 31 28](https://github.com/Expensify/App/assets/16821682/541b8141-0455-4bfc-af9a-ba3eec470176)

Also thx for the review, I wasn't pushing the latest changes 😅 now everything looks correct to me

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into perf/21022-sidebarlinks-performance

WoLewicki commented on src/libs/ReportUtils.js:
```diff
@@ -1916,13 +1915,6 @@ function shouldReportBeInOptionList(report, currentReportId, isInGSDMode, iouRep
         return false;
     }
 
-    // Include the currently viewed report. If we excluded the currently viewed report, then there
```
Are you sure we can just remove this condition and it will work correctly?

hannojg replied:
Haha okay, I may spoke too fast. I see that some tests are failing, indicating the need for this code. Will double and triple check 😊

hannojg replied:
Yeah, we can't remove that, otherwise the sorting in GSD/focus mode would break. I added it back but changed the general structure a bit to still be optimised.

mountiny commented:
@hannojg could we get back to this one soon? thanks 🙇

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into perf/21022-sidebarlinks-performance

hannojg committed:
fix issues after merge

hannojg committed:
fix import issues

hannojg committed:
change back name

hannojg committed:
remove currentReportId from params

mountiny commented:
@hannojg is this ready for review?

hannojg commented:
No not yet, sorry, the tests are still failing. Working on it right now

hannojg committed:
wip: data wrapper component

- Introduced a data wrppaer component, which gets connected to onyx
and constructs the report IDs from all the data
- Only a stable ref of report IDs gets passed to the actual sidebarlinks

hannojg committed:
logging

hannojg committed:
wip: fixing tests

hannojg commented:
Okay, this needed a bit of a larger rework, as we do still need the `currentReportID`. Without it we'd cause a regression in GSD mode (which I was overlooking previously).
I am however very happy with the rework, as I can already feel how the code gets cleaner, easier to understand and more efficient.

Will share details when finishing this tomorrow ✌️

hannojg committed:
add currentReportID back

hannojg committed:
wip(test): simpler context usage

hannojg committed:
wip: tests

hannojg committed:
wip: tests

hannojg committed:
include lastReadTime in wrapper component as its needed to compute GSD mode list correctly

hannojg committed:
fiz style in focus mode

hannojg committed:
fix SidebarFilterTests

hannojg committed:
wip: fixed all tests temporarily

hannojg committed:
removed focused index as its obsolete

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into perf/21022-sidebarlinks-performance

hannojg committed:
removed logging statements

hannojg committed:
split up components + use deepEqual in react memo

hannojg committed:
fix prop types

hannojg committed:
cleaned up tests

hannojg committed:
fix prop types

hannojg committed:
fixed last onyx data connections

hannojg committed:
clean

hannojg committed:
wip: moving comment logic over to data component

hannojg committed:
fix option row not upating when personal details changing

hannojg commented:
Okay, this is good for review again. Let me explain the changes I made on a high level to help with the code review:

## Splitting OptionRowLHN and SidebarLinks into UI and data components

There were two new components introduced:

- `<SidebarLinksData />`
- `<OptionRowLHNData />`

The corresponding none-data-components are now memoized, and are only expecting to receive derived values that are necessary for displaying the UI (and nothing more).

So, for example, `SidebarLinksData` calculates the list of report IDs we want to display given all the onyx data. Only when the calculated array really changed in its content (e.g. item order as in one report a new unread message arrived) it will pass the updated value to `SidebarLinks` causing an update.

The same goes for `OptionRowLHN` and its `optionItem` it wants to render. 

## Passing data from the components to onyx

Before, we were loading all the data needed for calculating the derived values for the components manually with `Onyx.connect` in e.g. `SidebarUtils`.
We also still had to connect the component to the onyx keys, to cause a re-render, which would then call the function to calculate the derived value (see e.g. `SidebarUtils.getOrderedReportIDs`).
This structure has its flaws, for example we were often missing to connect the right keys to the component, which would cause us to show stale data (See for example [this issue](https://github.com/Expensify/App/issues/21497))

We now connect all data needed to the component, and the component then calls the function with the data from the props. This is a transparent process that is easy to understand and reason about.
As this happens in the data component, we don't need to worry about unnecessary UI updates, as the underlaying UI component won't re-render unless the derived data really changed and we'd need to re-render.

### Future Considerations

The end game would be to not call the calculation function for derived data from the component using props, but from an onyx selector. In onyx we currently lack the functionality to have "dependent selectors".

For getting the user's personal data, we first need to load the `session` from onyx to get the user's `accountID` and then look into `personalDetailsList` (for this example, we currently have the `withCurrentUserPersonalDetails` hoc, which loads the `session` and `personalDetailsList` separately.)

Imagine we'd have the following in onyx:

```js
withOnyx({
   currentUserPersonalDetails: {
     keys: [ONYXKEYS.SESSION, ONYXKEYS.PERSONAL_DETAILS_LIST],
     selector: (session, personalDetails) => personalDetails[session.accountID]
   }
})
```

or another API could look like this:

```js
const currentUserPersonalDetailsSelector = createSelector(
   selectSessionAccountID,
   selectPersonalDetailsList,
   (accountID, personalDetails) => personalDetails[session.accountID]
)

withOnyx({
   currentUserPersonalDetails: currentUserPersonalDetailsSelector
})
```

The last example would even be better, as we don't rerun the selector when anything on the session changes, but just when the account ID changes.
I will try to make a proposal for this soon and then revisit.

## Utility `onyxSubscribe`

We want to avoid re-rendering our components for getting a new prop value, that is not needed for the UI output.
Earlier, we were connecting the `SidebarLinks` component to the `modal` key to get `willAlertModalBecomeVisible` which we needed for some keyboard shortcuts. 

`onyxSubscribe` is a convenience wrapper around `Onyx.connect` so we can read from onyx and receives updates, but not in the react state to avoid re-renders.

mountiny commented:
@fedirjh This is another one which could be higher on your list.

thanks @hannojg for pushing this forwards and great write up ❤️

WoLewicki commented on tests/unit/SidebarTest.js:
```diff
@@ -31,7 +32,10 @@ describe('Sidebar', () => {
     );
 
     // Initialize the network key for OfflineWithFeedback
```
Can you extend the comment to mention all things we do here?

WoLewicki commented on src/libs/SidebarUtils.js:
```diff
@@ -13,12 +13,16 @@ import * as CollectionUtils from './CollectionUtils';
-// There are a couple of keys here which are OK to have stale data. Having redundant subscriptions causes more re-renders which should be avoided.
-// Session also can remain stale because the only way for the current user to change is to sign out and sign in, which would clear out all the Onyx
-// data anyway and cause SidebarLinks to rerender.
+// Note: Earlier SidebarUtils.getOrderedReportIDs() used to have to parameters. All the needed data was loaded here directly
```
`used to have to parameters` what were you trying to say here?

WoLewicki reviewed (commented):
Added some comments. The PR looks good and I believe it implements the solutions we had in mind too 🚀 Good job! cc @j-piasecki did we have any other things that could be made better?

hannojg committed:
Update src/libs/SidebarUtils.js

Co-authored-by: Wojciech Lewicki <wojciech.lewicki@swmansion.com>

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into perf/21022-sidebarlinks-performance

hannojg committed:
fix issues after merge

hannojg committed:
dont use this

hannojg committed:
removed unused imports

hannojg committed:
fix comment

hannojg committed:
commented tests

hannojg commented on src/pages/home/sidebar/SidebarLinks.js:
```diff
@@ -253,13 +217,12 @@ class SidebarLinks extends React.Component {
                 </View>
-                {isLoading ? (
+                {this.props.isLoading ? (
                     skeletonPlaceholder
```
You are right, thx, no clue why it has been defined outside

hannojg committed:
inline skeleton

hannojg commented:
I incorporated the latest feedback and resolved conflicts. Eagerly waiting for your reviews 😊

fedirjh reviewed (commented):
Code looks solid. We should use props distrusting for functional components.

hannojg commented on src/pages/home/sidebar/SidebarLinks.js:
```diff
@@ -136,11 +91,19 @@ class SidebarLinks extends React.Component {
         SidebarUtils.setIsSidebarLoadedReady();
         this.isSidebarLoaded = true;
 
+        let modal = {};
+        this.unsubscribeOnyxModal = onyxSubscribe({
+            key: ONYXKEYS.MODAL,
+            callback: (modalArg) => {
+                modal = modalArg;
+            },
+        });
```
This is one of the performance improvements. The modal state is needed for a logic, that has nothing to do with the render/UI output. Thus, the component shouldn't re-render (because it's a really expensive to render one) just because we open a modal.

Long term I want to remove the modal state from onyx and move it e.g. to a react context, as I think it doesn't belong in the onyx store.

hannojg commented:
Thanks for the review @fedirjh will address all points/bugs shortly

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into perf/21022-sidebarlinks-performance

hannojg committed:
destructure props

hannojg commented on src/pages/home/sidebar/SidebarLinks.js:
```diff
@@ -275,122 +236,5 @@ class SidebarLinks extends React.Component {
-        },
-    }),
-)(SidebarLinks);
+export default compose(withLocalize, withCurrentUserPersonalDetails, withNavigationFocus, withWindowDimensions, withNavigation)(SidebarLinks);
```
It is used here:

https://github.com/margelo/expensify-app-fork/blob/5c35629da75b5fdca327c445c1f88aa392faff90/src/pages/home/sidebar/SidebarLinks.js#L154

and here

https://github.com/margelo/expensify-app-fork/blob/5c35629da75b5fdca327c445c1f88aa392faff90/src/pages/home/sidebar/SidebarLinks.js#L84-86

Although I think (in a follow up PR), we should be able to clean this up as well).

hannojg committed:
fix highlight active chat

hannojg committed:
add  displayName and iouReportAmount for sorting

hannojg committed:
include total, remove displayName

these were computed properties, iouReportAmount is computed,
same as displayname

hannojg committed:
added other properties considered for sorting

hannojg commented:
@fedirjh I addressed your comments and fixed all named bugs! Good for testing again

mountiny commented on src/pages/home/sidebar/SidebarLinksData.js:
```diff
@@ -0,0 +1,177 @@
+function SidebarLinksData({allReportActions, betas, chatReports, currentReportID, insets, isPersonalDetailsLoading, isSmallScreenWidth, onLayout, onLinkClick, policies, priorityMode}) {
+    const reportIDsRef = useRef([]);
+    const optionListItems = useMemo(() => {
+        const reportIDs = SidebarUtils.getOrderedReportIDs(currentReportID, chatReports, betas, policies, priorityMode, allReportActions);
```
@hannojg was this one addressed?

hannojg replied:
Whoops, sorry, missed to comment here.
Check the comment I wrote here: https://github.com/Expensify/App/pull/21406#discussion_r1265226360

That won't work. The `useMemo` only compares values by reference, not by its real content.

fedirjh reviewed (commented):
@hannojg bugs were fixed 👍🏼. I left some comments that should be resolved.

hannojg commented on src/pages/home/sidebar/SidebarLinksData.js:
```diff
@@ -0,0 +1,177 @@
+    const reportIDsRef = useRef([]);
+    const optionListItems = useMemo(() => {
+        const reportIDs = SidebarUtils.getOrderedReportIDs(currentReportID, chatReports, betas, policies, priorityMode, allReportActions);
+        if (deepEqual(reportIDsRef.current, reportIDs)) {
+            return reportIDsRef.current;
+        }
+        reportIDsRef.current = reportIDs;
```
That's because even when we put the same values to `SidebarUtils.getOrderedReportIDs` it will always return a new array reference. The same goes for every selector. So it can happen that one of the prop updates, but the real content of it stayed the same. In this case the `useMemo` will run and re-calculate the list of report IDs. The list will be exactly the same, but it's a new ref.If we'd return the new ref it would cause a re-render of `SidebarLinks`.
And we always must return from the `useMemo` hook. That's why we store the results in a ref, so in case its the very same result we can return the same reference.

hannojg committed:
use prop spreaing

hannojg commented on src/components/LHNOptionsList/OptionRowLHNData.js:
```diff
@@ -0,0 +1,132 @@
+ * The OptionRowLHN component is memoized, so it will only
+ * re-render if the data really changed.
+ */
+function OptionRowLHNData(props) {
```
Done 😊

hannojg committed:
move navigation focus to sidebarlinksdata

hannojg commented:
Hey, I added one other performance improvement. Instead of checking for the navigation focus to disable accessibility elements in `SidebarLinks`, I am doing that now in `SidebarLinksData`. With this change we prevent re-rendering the whole `FlatList` just because the `isFocused` prop changed (the propagation will stop at `SidebarLinks` and only `SidebarLinksData` will re-render)

hannojg committed:
removed debug comment

mountiny commented:
Thanks Hanno! 

@fedirjh seems like this is ready for another look 🚀

WoLewicki commented:
@hannojg I think we can remove `accessibilityElementsHidden` prop since it was needed when we did not have `stack` based navigation and now it is used only for tests (so we'd have to use something else there), but I am not 100% sure.  Code in stack for it: https://github.com/react-navigation/react-navigation/blob/24c03924397a6e59aba9f6b74a9c5cb4b939d9e1/packages/stack/src/views/Stack/CardContainer.tsx#L242.

hannojg commented:
Oh, would be even better to get rid of it completely @WoLewicki! Can you somehow confirm whether it can be removed or not?

@fedirjh You're the best tester, thank you so much for catching these! Will fix it shortly!

hannojg commented:
@fedirjh The `@expensify.sms` seems to be a general bug, I also have it on staging 👀 

![Screenshot 2023-07-17 at 18 14 20](https://github.com/Expensify/App/assets/16821682/15b92831-62de-4134-874c-5944cb11c684)

Once refreshed its gone

WoLewicki commented on src/pages/home/sidebar/SidebarLinks.js:
```diff
@@ -275,122 +231,5 @@ class SidebarLinks extends React.Component {
-        },
-    }),
-)(SidebarLinks);
+export default compose(withLocalize, withCurrentUserPersonalDetails, withWindowDimensions, withNavigation)(SidebarLinks);
```
We are currently investigating another issue and it seems that `withNavigation` is no longer used in that component. Or are we missing something?

hannojg replied:
@WoLewicki i think you are right, good catch. There is the usage of `Navigation.js` which uses the global navigation ref. I am not sure if `useNavigation` is there to make sure the navigationRef is ready, but either way i think it can get removed

hannojg replied:
removed it 🥷

hannojg committed:
remove withNavigation usage

hannojg committed:
add other missing fields

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into perf/21022-sidebarlinks-performance

hannojg commented:
@fedirjh Would appreciate another review 🙏 (Sorry if there are still bugs, there are many conditions to the LHN and I really hope we catch them all! In my testings everything looks good, curious about yours)

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into perf/21022-sidebarlinks-performance

hannojg commented:
I might be wrong but i think localhost is using the development code version, which is less optimized and slower, while staging already uses an optimized build, that is closer to the final release build.
But it might be that it caused a performance regression in that sense, will look into it today!

hannojg committed:
Merge branch 'main' of github.com:margelo/expensify-app-fork into perf/21022-sidebarlinks-performance

hannojg commented:
@fedirjh Hey, this is a bit hard to reproduce for me. However, I have a theory of why this might be happening.

I added a patch for onyx, from this PR:

- https://github.com/Expensify/react-native-onyx/pull/280

it removes the blocking behaviour from `withOnyx`.

Can you please pull, run `npc patch-package`, make sure the onyx patch got applied, and try testing for performance again?
In addition it'd be cool if you could run `npm run build-staging` and then test with the release build.

Thank you so much in advance!

hannojg commented:
You can, or i think you can run the build command and then use something like [npx serve](https://www.npmjs.com/package/serve) to try out the build

hannojg commented:
Okay, i reverted the patch!

hannojg committed:
removed unused onLayout prop

hannojg committed:
Memoize OptionRowLHNData as well

hannojg commented:
I pushed two little last changes. I think we should be good now

hannojg committed:
fix lint

fedirjh reviewed (approved):
Looks good to me and tests well

mountiny commented on src/components/LHNOptionsList/OptionRowLHNData.js:
```diff
@@ -0,0 +1,140 @@
+    );
+
+// This component is rendered in a list.
+// On scroll we want to avoid that a item re-renders
```
```suggestion
// On scroll we want to avoid item re-renders
```

mountiny commented on src/components/LHNOptionsList/OptionRowLHNData.js:
```diff
@@ -0,0 +1,140 @@
+        {},
+    );
+
+// This component is rendered in a list.
```
Could you also use the `/ **` style for this, please?

mountiny commented on src/libs/ReportActionsUtils.js:
```diff
@@ -97,13 +97,14 @@ function hasCommentThread(reportAction) {
  * @returns {Object}
  */
-function getParentReportAction(report) {
+function getParentReportAction(report, allReportActionsParam) {
```
If its optional should we give it an empty object as a default value?
```suggestion
function getParentReportAction(report, allReportActionsParam = {}) {
```

mountiny reviewed (dismissed):
Great job on this one @hannojg and @fedirjh just couple of comments

WoLewicki reviewed (approved):
LGTM 🚀

hannojg committed:
Update src/components/LHNOptionsList/OptionRowLHNData.js

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

hannojg committed:
Update src/libs/ReportActionsUtils.js

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

hannojg committed:
change comment format

0xmiroslav reviewed (commented):
Some issues after logout and re-login:
- LHN items order changed (2nd and 3rd are swapped in video)
- go to any random report after login, instead of last opened report

https://github.com/Expensify/App/assets/97473779/f7a743c3-ff48-475c-859d-a63dd5d41035

mountiny reviewed (approved):
@hannojg  @fedirjh Lets be on a look out for deploy blockers tomorrow or after the next deploy. I know you have tested well but with such PRs its always tricky and high chance of something breaking would be great to have you around to follow up with a fix as reverting would be messy

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

mountiny commented:
Thanks was not sure if you are reviewing, sorry

hannojg commented:
@0xmiroslav looking into it now!

hannojg commented:
@0xmiroslav Oh, I have seen this as well and thought I'd be a bug! but its on staging as well. Pinned rooms are only sorted by their name. When two workspaces have the same name their order will always be random (often times its by luck that its the same, but its really random) 😅

hannojg commented:
https://github.com/Expensify/App/assets/16821682/91462cf2-c974-4616-9816-d5cabe798c44

Proof that I am no liar 😄

hannojg commented:
@mountiny is it possible that something broke during merge?

@0xmiroslav are you on the latest version of the perf branch?

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5602380038) to staging by https://github.com/mountiny in version: 1.3.43-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

mountiny commented:
@hannojg it is possible

jasperhuangg commented:
This PR has caused a number of regressions:

https://github.com/Expensify/App/issues/23189
https://github.com/Expensify/App/issues/23195
https://github.com/Expensify/App/issues/23202

I'm going to revert it

aldo-expensify commented on src/libs/ReportActionsUtils.js:
```diff
@@ -97,13 +97,14 @@ function hasCommentThread(reportAction) {
  * @returns {Object}
  */
-function getParentReportAction(report) {
+function getParentReportAction(report, allReportActionsParam = {}) {
```
This default is causing the deploy blocker here: https://github.com/Expensify/App/issues/23202

If `allReportActionsParam` is not passed, it has value `{}` which is truthy, so we never use `allReportActions`. The `allReportActionsParam || allReportActions` doesn't work as intended.

hannojg replied:
thanks you for quickly fixing this!!

mountiny replied:
Thanks for the fix

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5602380038) to staging by https://github.com/mountiny in version: 1.3.43-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5625378789) to staging by https://github.com/mountiny in version: 1.3.44-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5625371664) to production by https://github.com/jasperhuangg in version: 1.3.43-7 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5651460665) to production by https://github.com/marcaaron in version: 1.3.44-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

s77rt commented on src/pages/home/sidebar/SidebarLinksData.js:
```diff
@@ -0,0 +1,200 @@
+    withOnyx({
+        chatReports: {
+            key: ONYXKEYS.COLLECTION.REPORT,
+            selector: chatReportSelector,
```
This caused a "regression"? Without the use of the selector we used to pass the reports objects as is in Onyx. Then we mutated those objects here https://github.com/Expensify/App/blob/bba0eb2868fa95d483da1b380551bc46fe0df427/src/libs/SidebarUtils.js#L109-L110 by adding a `displayName` to each report. If you try to access `report.displayName` from anywhere in the App you'd be able to (because we modified the onyx data).

Now, we are using a selector and we are returning a copy of Onyx data, we are mutating a clone and not the onyx data. If we try to access `report.displayName` from anywhere outside the Sidebar we won't get a result which in turn caused this issue https://github.com/Expensify/App/issues/23348.

hannojg replied:
Wow great catch! But at the same time I have to say that how it was working before is very dirty and we should find a different solution for it 😅

hannojg replied:
I see that the proposal proposes a clean way, good job 💪

Santhosh-Sellavel commented:
This PR Causes [this](https://github.com/Expensify/App/issues/23676) regression more details [here](https://github.com/Expensify/App/issues/23676#issuecomment-1652400735)

marcaaron commented on tests/unit/SidebarFilterTest.js:
```diff
@@ -253,6 +258,7 @@ describe('Sidebar', () => {
             );
         });
 
+        // NOTE: This is also for #focus mode, should we move this test block?
```
I am not sure what the original intention here was. Maybe @tgolen can remember. This comment stuck out to me as pretty confusing and something we should fix.

**context:** Working on a PR and encountered a failing test with this title:

>Sidebar › in default (most recent) mode › all combinations of isArchived, isUserCreatedPolicyRoom, hasAddWorkspaceError, isUnread, isPinned, hasDraft › the booleans true,false,false,true,false,false

But when I look at the lines of code it's referencing I see that we are not testing "most recent" but "focus" mode.

marcaaron replied:
This test also confuses me if the intention is that we should be testing "Most recent" mode. 

If we are trying to see that a report is in the "Most recent" list then most all reports will be in the list.

Or at least doesn't seem dependent on the arguments we are passing.

>isArchived, isUserCreatedPolicyRoom, hasAddWorkspaceError, isUnread, isPinned, hasDraft

All of those will would show in "Most recent" (all other things about the report being "normal")

marcaaron replied:
Ok after reviewing the two blocks - I have determined they are identical and we should be able to remove this...

I don't think this was intended to be put here and it's not logical to test.
